### PR TITLE
make scpi port and events (connected, error) usable in project

### PR DESCRIPTION
**Pull request by MisterHW**
* move DEVICE_PORT and SERVER_PORT to header
* add weak event handlers for accepted / closed connections and error indication

After copying scpi-parser files into a project and working with the LWIP-netconn example for a few months, this is the first of perhaps a few PRs with bug fixes and enhancements introduced to an scpi-parser fork in a hopefully clean manner to offer them to the community.

I'm using the event handlers in question to turn on  / off user LEDs on a Nucleo-F767ZI board.

### CODE REVIEW

File: examples/test-LwIP-netconn/scpi_server.h [hunk 1]
```diff
@@ -28,14 +28,29 @@
...
+#include "api.h"
+#include "scpi/types.h"
 
 void scpi_server_init(void);
 
 void SCPI_AddError(int16_t err);
 void SCPI_RequestControl(void);
 
+// optional event handlers
+void SCPI_Event_DeviceConnected(struct netconn * conn);
```
Comment (j123b567) [inline examples/test-LwIP-netconn/scpi_server.h:48]: It would be great to have also `scpi_t * context` as the first parameter for all events.

File: examples/test-LwIP-netconn/scpi_server.c [hunk 2]
```diff
@@ -117,11 +113,10 @@ int SCPI_Error(scpi_t * context, int_fast16_t err) {
     iprintf("**ERROR: %ld, \"%s\"\r\n", (int32_t) err, SCPI_ErrorTranslate(err));
```
Comment (j123b567) [inline examples/test-LwIP-netconn/scpi_server.c:113]: In this sense, for me, it would be better to move iprintf to example error handlers.

File: examples/test-LwIP-netconn/scpi_server.c [hunk 3]
```diff
@@ -252,6 +247,7 @@ static int processIoListen(user_data_t * user_data) {
             netconn_delete(newconn);
         } else {
             /* connection established */
+            SCPI_Event_DeviceConnected(newconn);
             iprintf("***Connection established %s\r\n", inet_ntoa(newconn->pcb.ip->remote_ip));
```
Comment (j123b567) [inline examples/test-LwIP-netconn/scpi_server.c:251]: Also please move iprintf to the event handler

File: examples/test-LwIP-netconn/scpi_server.c [hunk 4]
```diff
@@ -152,7 +147,7 @@ scpi_result_t SCPI_Reset(scpi_t * context) {
 }
```
Comment (j123b567) [inline examples/test-LwIP-netconn/scpi_server.c:147]: In this context, `SCPI_Reset` is the same type of event so it would be also `weak`.

File: examples/test-LwIP-netconn/scpi_server.h [hunk 5]
```diff
@@ -28,14 +28,29 @@
...
 #define _SCPI_SERVER_H_
+#ifdef __cplusplus
+extern "C" {
+#endif
 
+#define SCPI_DEVICE_PORT  5025 // scpi-raw standard port
+#define SCPI_CONTROL_PORT 5026 // libscpi control port (not part of the standard)
 
 #include <stdint.h>
+#include "api.h"
```
Comment (j123b567) [inline examples/test-LwIP-netconn/scpi_server.h:39]: This file `api.h` is not part of the PR

**Review by j123b567 — CHANGES_REQUESTED** [review]
This was intended just as an example and not directly usable without modifications.
But further extension are welcome.

File: examples/test-LwIP-netconn/scpi_server.h (hunk 5 above)
Comment (MisterHW) [reply examples/test-LwIP-netconn/scpi_server.h:39]: This probably needs to be
```
#include "lwip/api.h"
```
for netconn to be defined.

Comment (j123b567) [reply examples/test-LwIP-netconn/scpi_server.h:39]: So, please, use `#include "lwip/api.h"` instead of bare `#include "api.h"`.

File: examples/test-LwIP-netconn/scpi_server.h (hunk 1 above)
Comment (MisterHW) [reply examples/test-LwIP-netconn/scpi_server.h]: realized in follow-up commit. It seems to me that &scpi_context would need to be passed n scpi_server_thread()  to toprocessIoListen(), processSrqIoListen(), processIo() and processSrqIo(). This requires the respective signatures to be extended. If that's what you implied, I can add it in another commit.

Comment (MisterHW) [reply examples/test-LwIP-netconn/scpi_server.h]: see a565c895080f17deda47062e00f76a27040e6c17

**Review by j123b567 — CHANGES_REQUESTED** [review]
One final change, can you please reformat the file so only spaces are used for indention?
After this, I don't see any other obstacle for merging.

**Review by j123b567 — APPROVED** [review]
